Add docs for raster datasets (#849)
koordinates/kart: docs/pages/development/raster_v1.rst

Raster Datasets V1
-----------------------

Overall structure
~~~~~~~~~~~~~~~~~

> A V1 raster dataset is a folder named ``.raster-dataset.v1`` that contains two folders. These are ``meta`` which contains information about the entire dataset, irrespective of a particular tile - and ``tile``, which contains raster tiles in the ``GeoTIFF`` format. The schema file contains information about each band that is stored for each grid-cell / pixel in the raster. The "name" of the dataset is the path to the ``.raster-dataset.v1`` folder.

Reviewer: ```suggestion
A V1 raster dataset is a folder named ``.raster-dataset.v1`` that contains two folders. These are ``meta`` which contains information about the entire dataset, irrespective of a particular tile - and ``tile``, which contains raster tiles in the ``GeoTIFF`` format. The schema file contains information about each band in the raster. The "name" of the dataset is the path to the ``.raster-dataset.v1`` folder.
```

``aerials/north-head``:

::

   aerials/
   aerials/north-head/
   aerials/north-head/.raster-dataset.v1/
   aerials/north-head/.raster-dataset.v1/meta/
   aerials/north-head/.raster-dataset.v1/meta/title              # Title of the dataset
   aerials/north-head/.raster-dataset.v1/meta/description        # Description of the dataset
   aerials/north-head/.raster-dataset.v1/meta/format.json        # Format of the dataset
   aerials/north-head/.raster-dataset.v1/meta/schema.json        # Schema of the dataset
   aerials/north-head/.raster-dataset.v1/meta/crs.wkt            # CRS of the dataset
   aerials/north-head/.raster-dataset.v1/meta/band-1/...         # Extra metadata specific to the first band...
   aerials/north-head/.raster-dataset.v1/meta/band-2/...         # Extra metadata specific to the second band...
   aerials/north-head/.raster-dataset.v1/meta/...                # Other dataset metadata

   aerials/north-head/.raster-dataset.v1/tile/...                # All of the tiles in the dataset - all GeoTIFFs

Meta items
~~~~~~~~~~

The following items are stored in the meta part of the dataset, and have the following structure.

``meta/title``
^^^^^^^^^^^^^^

> Contains the title of the dataset, encoded using UTF-8. The title is freeform text, clients could if they desired include HTML tags or markdown in the title.

Reviewer: ```suggestion
Contains the title of the dataset, encoded using UTF-8. The title is freeform text.
```
I don't think that last clause adds much.

``meta/description``
^^^^^^^^^^^^^^^^^^^^

> Contains the title of the dataset, encoded using UTF-8. The description is freeform text, clients could if they desired include HTML tags or markdown in the title.

Reviewer: ```suggestion
A long-form description of the dataset, encoded using UTF-8. The description is freeform text and has no enforced length limit.
```

``meta/format.json``
^^^^^^^^^^^^^^^^^^^^

Contains information about constraints related to file format that the GeoTIFF tiles must follow. A single dataset cannot contain a mixture of all kinds of GeoTIFF files - there must be some commonality between them. For example, every tile in a dataset must have the same CRS.

In terms of format, currently there are only one or two constraints stored in the format.json that every tile in the dataset must follow:
1. All tiles must be GeoTIFF (this constraint is always present for Raster Datasets V1 - no other file types are supported).
2. All tiles must be `Cloud-Optimized GeoTIFFs <cog_>`_. This constraint is only set for certain datasets where the user has requested it.

Constraining the tiles to be Cloud Optimized GeoTIFFs (COGs) means they can be viewed by webviewers - the COG format is a subtype of GeoTIFF file which allows for fasting seeking and rendering of whichever part of the raster is currently within the user's viewport.

For example, here is the format of a dataset that has both constraints:

.. code:: json

    {
      "fileType": "geotiff",
      "profile": "cloud-optimized"
    }

> If the cloud-optimized constraint is relaxed, such that both COGs and non-COGs are allowed, then the relevant field is ommitted.

Reviewer: ```suggestion
If the cloud-optimized constraint is relaxed, such that both COGs and non-COGs are allowed, then the `profile` field is omitted.
```

.. _raster-meta-schema-json:

``meta/schema.json``
^^^^^^^^^^^^^^^^^^^^

> Contains information about the "bands" of information, that is, each unit of data that is stored per-pixel. This item has a similar format as in other types of datasets, but one important difference is that the each band in a raster dataset has no identifying ID or name - the properties that identify a raster band are the following:

Reviewer: ```suggestion
Contains information about the "bands" of information, that is, each unit of data that is stored per-pixel. This item has a similar format as in other types of datasets, but one important difference is that the bands in a raster dataset have no identifying ID or name. The properties that identify a raster band are the following:
```

> 1. Its position in the list.

Reviewer: ```suggestion
1. It's position in the list.
```

Author: Not done - "its" is actually correct.
https://www.merriam-webster.com/words-at-play/when-to-use-its-vs-its

Reviewer: ok - this one's pretty confusing though hey
maybe fix line 74 then?

Author: Good point. I can only see snippets of what I wrote in this review-conversation so can't see the line right next to it.

Author: Done

This XML contains the column headings of the `Raster Attribute Table <rat_>`_ associated with this band, if there is one. This information is extracted from the `Persistent Auxiliary Metadata <pam_>`_ (PAM) files (the ``.aux.xml`` files) associated with the tiles, if there are any.

If a particular band has a raster-attribute-table, it must have the same raster attribute table for every tile: that is, every instance of that raster attribute table attached to each tile should have the same number of fields with the same field definitions. However, the rows of the raster attribute table may vary from tile to tile. This is a continuation of the principle that every tile should have the same schema, but the actual data of each tile may vary to any extent.

``meta/band/1/categories.json``
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

This is indexed starting at one - ``band/1/categories.json`` is for the first band, ``band/2/categories.json`` is for the second band, and so on. Only certain bands will have this metadata attached.

This JSON object describes a mapping from each possible value that the band can take, to a freeform text description of what that value means. As a simple example,

> it could be ``{"1": "Land", "2": "Water"}``. This information is extracted from the `Persistent Auxiliary Metadata <pam_>`_ (PAM) files (the ``.aux.xml`` files)associated with the tiles, if there are any.

Reviewer: ```suggestion
it could be ``{"1": "Land", "2": "Water"}``. This information is extracted from the `Persistent Auxiliary Metadata <pam_>`_ (PAM) files (the ``.aux.xml`` files) associated with the tiles, if there are any.
```
